### ISSUE TITLE
Refactor TestCase classes

### ISSUE THREAD

**Issue by Marenz** [ethereum/solidity#8244]
@ekpyron and I went a bit over the testcases code (while looking at #8238) and come up with the following ideas for a refactor:

* remove `validateSettings` and the related variables `m_settings` and `m_validateSettings`. Instead add a function `void addSetting()` that takes a string and a lambda/function pointer that is called for each setting that is found. The lambda can set desired member variables or throw if the setting is invalid.
* To keep the mechanic of `validateSettings` that we can cause tests not to be run based on various conditions the idea is to add a member variable `bool m_shouldRun = true` that any member function can set to false
* There is currently logic in both `YulOptimizerTest` and `libyul/SyntaxTest` that validates/handles the `dialect` option that should be unified.
* the name of `printExpectationAndError` does not make it obvious that it checks for errors, too

**Comment by Marenz** [comment]
@a3d4 started working on this

**Comment by a3d4** [comment]
If we introduce `addSetting()`, the following issue has to be addressed.

Consider, for instance, `SMTCheckerTest` and `SyntaxTest`. Both classes can be instantiated and `SMTCheckerTest` is derived from `SyntaxTest`. (That violates Meyer's "Make non-leaf classes abstract" recommendation BTW.)

`SMTCheckerTest` has `"SMTSolvers"` setting. `SyntaxTest` has `"optimize-yul"` setting.

If we construct `SyntaxTest` as a base class of `SMTCheckerTest`, then parsing of settings should not happen (it should be done later during construction of `SMTCheckerTest`).

But if we construct `SyntaxTest` on its own, then parsing of settings should happen.

There are a few options to implement it, but none is particularly appealing:
**[a]** change class hierarchy to allow instantiation of most-derived classes only;
**[b]** add a second constructor to SyntaxCheck and use different constructors for different scenarios;
**[c]** add `Init()` function to both classes, keep `addSetting()` in the constructors, move setting parsing to `Init()`, and call `Init()` before returning an objects from `SyntaxTest::create()` and `SMTCheckerTest::create()`.

I would consider **[c]** as a smaller evil (combined with making the constructors protected).

However, it might be significantly simpler just

**[d]** to keep `m_settings` in the base class, read all available settings during construction of the base class, and then allow derived classes to convert their settings in their constructors in an explicit way, something like `m_optimizeYul = initBoolSetting("optimize-yul")`.

There are not so many settings, so **[d]** looks less invasive. And the classes would have all their settings (and `m_shouldRun` flag) initialized already during construction.

What do you think?

**Comment by Marenz** [comment]
for solution **[d]**, how would that work with non bool/int settings like evm version?

**Comment by a3d4** [comment]
It could be a static utility function, e.g. `initEvmVersionSetting()`. Or we can handle it the same way as now. Since we have a limited number of settings, the first option might be slightly better.

**Comment by Marenz** [comment]
So one thing we can consider is that there is a `create` function for all the classes that takes care of init. Maybe we can add a call to the `init()` function in there and make the c'tor protected?

**Comment by a3d4** [comment]
> There is currently logic in both YulOptimizerTest and libyul/SyntaxTest that validates/handles the dialect option that should be unified.

There is a subtle difference. For `"evm"` we use 
`EVMDialect::strictAssemblyForEVMObjects` in `YulOptimizerTest` and 
`EVMDialect::strictAssemblyForEVM` in `SyntaxTest`
```
else if (dialectName == "evm")
    m_dialect = &EVMDialect::strictAssemblyForEVMObjects( . . . . );
                                                 ^^^^^^
```

```
"evm",
[](langutil::EVMVersion _evmVersion) -> yul::Dialect const&
{ return yul::EVMDialect::strictAssemblyForEVM(_evmVersion); }
```

With `strictAssemblyForEVMObjects` in both cases, tests pass. With `strictAssemblyForEVM`, some tests fail.

What is the intended behavior?

**Comment by chriseth** [comment]
Please use `...EVMObjects`, it is more generic than the other one.

**Comment by a3d4** [comment]
Proposed draft is PR #8442.

I did consider reading settings via callbacks (`addSetting()` + lambda). This approach looks elegant, but would result in (IMO) more complex and repetitive code.

* We could not keep the logic in ctors only. Might need to move ctor logic to somewhere like virtual `init()`, and call it from every `create()`, as @Marenz mentioned. (The reason is that we cannot call lambda for subclass settings from a superclass ctor.)
* Interface-wise, it is not ideal to get source(s) and expectation(s) as explicit function results, but settings as a side effect of the function.
* Via lambdas, we can only set settings if their values are present in the file. The default values have to be assigned to settings somewhere else. So initialization of a setting has to spread across two places.
* For `TestCase::printUpdatedSettings` we do need one container with all settings anyway.

So I would propose direct reading of everything in ctors.

I introduced `TestCaseReader` class which reads source(s), settings and simple expectations. It is not ideal either.

* Custom expectation readers belong to corresponding `TestCase` subclasses. Such subclasses read from the stream, partly parsed by TestCaseReader. It is eclectic.

But all in all, it looks as a smaller evil to me. Let's discuss :).
